# TGS Gap Filler

[![DOI](https://zenodo.org/badge/183120917.svg)](https://zenodo.org/badge/latestdoi/183120917)

A gap filling tool that uses long reads generated by thild-generation-sequence techniques (Pacbio, Oxford nanopore, etc) to fill N-gap in the genome assembly.

- Both raw reads and error-corrected reads can be used as input .

- If additional NGS short reads are available, it polishes raw TGS reads by calling Pilon .

## How to install

### Download 
```
git clone  https://github.com/BGI-Qingdao/TGSGapFiller.git YOUR-INSTALL-DIR
```

### Compile

#### prepare minimap2

- If you already have a minimap2 installed, you can link it here

```
rm -rf YOUR-INSTALL-DIR/minimap2
ln -s MINIMAP2-PATH YOUR-INSTALL-DIR/
```
- Otherwise you need to download it by :

```
cd YOUR-INSTALL-DIR
git submodule init
git submodule update
```

#### compile

```
cd YOUR-INSTALL-DIR/main_src_ont
make
```

## Usage 

### basic usage 

```
Usage:
      TGSGapFiller --scaff SCAFF_FILE --reads TGS_READS_FILE --output OUT_PREFIX [options...]
      required :
          --scaff     <scaffold_file>      the input scaffold file.
          --reads     <tgs_reads_file>     the input TGS reads file.
          --output    <output_prefix>      the output prefix.
      part required :
          --ne                             do not error correct. error correct by default.
          or
          --racon     <racon>              the installed racon.
          or
          --ngs       <ngs_reads>          the ngs reads used for pilon
          --pilon     <pilon>              the installed pilon.
          --samtool   <samtool>            the installed samtool.
          --java      <java>               the installed java.
      optional:
          --tgstype   <pb/ont>             TGS type . ont by default.
          --min_idy   <min_idy>            min_idy for filter reads .
                                           0.3 for ont by default.
                                           0.2 for pb by default.
          --min_match <min_idy>            min match length for filter reads .
                                           300bp for ont by default.
                                           200bp for pb by default.
          --thread    <t_num>              thread uesd . 16 by default.
          --pilon_mem <t_num>              memory used for pilon , 300G for default.
          --chunk     <chunk_num>          split candidate into chunks to error-correct separately.
```

### version without error correction 

* example

```
YOUR-INSTALL-DIR/TGSGapFiller.sh  \
        --scaff  scaffold-path/scaffold.fasta \
        --reads  tgs-reads-path/tgs.reads.fasta \
        --output test_ne  \
        --ne \
        >pipe.log 2>pipe.err
```

### version with error correction by racon

```
YOUR-INSTALL-DIR/TGSGapFiller.sh  \
        --scaff  scaffold-path/scaffold.fasta \
        --reads  tgs-reads-path/tgs.reads.fasta \
        --output test_racon \
        --racon  raconn-path/bin/racon \
        >pipe.log 2>pipe.err
```

### version with error correction by pilon

```
YOUR-INSTALL-DIR/TGSGapFiller.sh  \
        --scaff  scaffold-path/scaffold.fasta \
        --reads  tgs-reads-path/tgs.reads.fasta \
        --output test_pilon \
        --pilon  pilon-path/pilon-1.23.jar  \
        --ngs    ngs-reads-path/ngs.reads.fastq  \
        --samtool samtools-path/bin/samtools  \
        --java    java-path/bin/java \
        >pipe.log 2>pipe.err
```

### deal with pacbio input reads

* default thild-generation-reads type is ont , use ```--tgstype```  to change it .

```
    --tgstype ont
    or 
    --tgstype pb
```

* An example of use pacbio input reads and use racon to do error-correction

```
YOUR-INSTALL-DIR/TGSGapFiller.sh  \
        --scaff  scaffold-path/scaffold.fasta \
        --reads  tgs-reads-path/tgs.reads.fasta \
        --output test_racon \
        --racon  raconn-path/bin/racon \
        --tgstype pb \
        >pipe.log 2>pipe.err
```

## Output

- your-prefix.scaff_seq 
    - this is the final sequence after gap filling
- your-prefix.gap_fill_details
    - details about how the final sequence was assemblied 

### format of your-prefix.gap_fill_details

#### a example :

```
>scaffold_1
1	1000	S	1000	2000
1001	1010	N
1011	1100	S	2201	2290
1101	1110	F
1111	1200	S	2301	2390
>scaffold_2
......

```
#### detailed information

1. each scaffold name is followed by its data lines.
2. data line contains 3 or 5 columns and describes the source of a segments of the final sequence:
    - column 1 is the segment's first bp position at the final sequence.
    - column 2 is the segment's last bp position at the final sequence.
    - column 3 is the segment's type , 'S' , 'N' or 'F'.
        - 'S' means this segment is a segment of the input sequence and this line includes other 2 more columns :
            - column 4 is the segment's first bp position at the input sequence.
            - column 5 is the segment's last bp position at the input sequence.
        - 'N' means this segment is N area.
        - 'F' means this segment is a filled sequence from TGS reads .

